### ISSUE TITLE
core: large sampling rates do not play well with remembering past digital I/O and analog out values 

### ISSUE THREAD

**Issue by giuliomoro** [BelaPlatform/Bela#604]
This is because "persistency" is implemented in `PRU::loop()`, while it should happen around wherever the user callback is called.

**Comment by giuliomoro** [comment]
sorry for the late response. You should  call `digitalWrite()` at least once per block and additionally once per block you should also call

```
pinMode(context, 0, gOutputPin, OUTPUT)
```
(where `gOutputPin` is the pin you want to set as an output)

**Comment by akiomik** [comment]
I reproduced this on a Gem Stereo with the Bela Debian Bookworm Image 2026-03-25.

The ordinary digital loopback probe configures its output once, then writes only once every eight application blocks:

- period 255: one loopback edge per scheduled write;
- period 256: no loopback edges and no output LED changes;
- period 320: no loopback edges and no output LED changes.

At 256 and 320, audio continues normally and `Bela_initAudio()` succeeds. The application callback is running through `BelaContextFifo` at those periods.

I then applied the workaround from this issue: on every application block, re-apply the output `pinMode(..., 0, ..., OUTPUT)` and the current digital value from frame 0, while keeping the scheduled transition at frame 4. That restores one loopback edge per scheduled write at periods 256 and 320, with no misses or unexpected edges. The 255-frame result is unchanged.

I also tested D1 independently of a Bela digital output: a Linux-controlled GPIO drove D1 through 1 kΩ. At periods 255, 256, and 320, each of thirteen consecutive one-second windows contained exactly one D1 edge, alternating level. This rules out complete loss of that measured input path; the failure is consistent with output/direction persistence being placed outside the application callback.

The probes and measurement details are available in:
- https://github.com/akiomik/bela-rs/pull/106
- https://github.com/akiomik/bela-rs/pull/107

Could the persistence of digital directions and values be moved around the user callback when it runs through `BelaContextFifo`? I can retest the Gem Stereo runs against an upstream fix.

**Comment by giuliomoro** [comment]
Yeah I never fixed this. The workaround is to call pinMode()/digitalWrite()/analogWrite() once per block or  pinModeOnce()/digitalWriteOnce()/analogWriteOnce() once per frame.
